fedimint-api-client/src/api/global_api/with_online_toggle.rs

```diff
@@ -0,0 +1,95 @@
+use std::collections::BTreeSet;
+
+use anyhow::format_err;
+use fedimint_core::core::ModuleInstanceId;
+use fedimint_core::module::ApiRequestErased;
+use fedimint_core::task::{MaybeSend, MaybeSync};
+use fedimint_core::{apply, async_trait_maybe_send, PeerId};
+use serde_json::Value;
+use tokio::sync::{watch, RwLock};
+
+use super::super::{DynModuleApi, IRawFederationApi};
+use crate::api::{PeerError, PeerResult};
+
+/// Convenience extension trait used for wrapping [`IRawFederationApi`] in
+/// a [`RawFederationApiWithOnlineToggle`]
+pub trait RawFederationApiWithOnlineToggleExt
+where
+    Self: Sized,
+{
+    fn with_online_toggle(
+        self,
+        online: watch::Receiver<bool>,
+    ) -> RawFederationApiWithOnlineToggle<Self>;
+}
+
+impl<T> RawFederationApiWithOnlineToggleExt for T
+where
+    T: IRawFederationApi + MaybeSend + MaybeSync + 'static,
+{
+    fn with_online_toggle(
+        self,
+        online: watch::Receiver<bool>,
+    ) -> RawFederationApiWithOnlineToggle<T> {
+        RawFederationApiWithOnlineToggle::new(self, online)
+    }
+}
+
+/// [`IRawFederationApi`] wrapping some `T: IRawFederationApi` and adding
+/// a toggle for online/offline state.
+///
+/// If the state is offline, all requests will block until toggle switched to
+/// online, without actually making any connections.
+///
+/// Use [`RawFederationApiWithOnlineToggleExt::with_online_toggle`] to
+/// create.
+#[derive(Debug)]
+pub struct RawFederationApiWithOnlineToggle<T> {
+    pub(crate) inner: T,
+    online: tokio::sync::RwLock<watch::Receiver<bool>>,
+}
+
+impl<T> RawFederationApiWithOnlineToggle<T> {
+    pub fn new(inner: T, online: watch::Receiver<bool>) -> RawFederationApiWithOnlineToggle<T> {
+        Self {
+            inner,
+            online: RwLock::new(online),
+        }
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+impl<T> IRawFederationApi for RawFederationApiWithOnlineToggle<T>
+where
+    T: IRawFederationApi + MaybeSend + MaybeSync + 'static,
+{
+    fn all_peers(&self) -> &BTreeSet<PeerId> {
+        self.inner.all_peers()
+    }
+
+    fn self_peer(&self) -> Option<PeerId> {
+        self.inner.self_peer()
+    }
+
+    fn with_module(&self, id: ModuleInstanceId) -> DynModuleApi {
+        self.inner.with_module(id)
+    }
+
+    async fn request_raw(
+        &self,
+        peer_id: PeerId,
+        method: &str,
+        params: &ApiRequestErased,
+    ) -> PeerResult<Value> {
+        if !*self.online.write().await.borrow() {
+            return Err(PeerError::Connection(format_err!("Online toggle is off")));
+        }
+        // Note: We could block until online too...
```

Comment:
I'm a bit worried that failing connection right away will just lead to aggressive retry loops.

We could either: block here, or block with a timeout.

Or add some `fn is_fatal` to `PeerError` and learn all retry loops to use it ...

Reply:
IMO just sleep for some time (not wait for online)

Reply:
That might block indefinitely, preventing something in the higher level from giving up if it isn't using a custom channel.


How about wait for notification, but wrap in a 60s timeout and return `Err(PeerError::Connection(format_err!("Online toggle is off")))` on timeout? @maan2003

Reply:
> That might block indefinitely, preventing something in the higher level from giving up if it isn't using a custom timeout. Which most things should have (timeout), but we might have forgotten somewhere, and a normal connection would eventually bubble up anyway.

this is good if we find such cases, IMO job of this toggle is to find bugs